src/java/htsjdk/samtools/SamPairUtil.java

```diff
@@ -466,6 +467,7 @@ private void advance() {
         }
 
         public SAMRecord next() {
+            if(!hasNext()) { throw new NoSuchElementException(); }

```

Comment:
Hi Yossi,

I’m not sure what assertion I added.  Can you clarify?

Having next() throw NoSuchElementException is merely conforming to the java.util.Iterator.next() contract.  Are you saying that PeekableIterator.hasNext() throws an exception if the iterator is closed?  I don’t see where that happens.  This is all I see for PeekableIterator.hasNext():


    public boolean hasNext() {
        return this.nextObject != null;
    }

I don’t see how this would throw an exception.  

Anyway, the CloseableIterator interface doesn’t specify what should happen when you call next() or hasNext() on a closed iterator.  IMHO, I think throwing an exception when a call is made on a closed iterator (other than closing again) seems like a desirable behavior.

-Alec

> On Feb 19, 2016, at 9:34 PM, Yossi Farjoun <notifications@github.com> wrote:
> 
> In src/java/htsjdk/samtools/SamPairUtil.java <https://github.com/samtools/htsjdk/pull/480#discussion_r53542692>:
> 
> > @@ -466,6 +467,7 @@ private void advance() {
> >          }
> >  
> >          public SAMRecord next() {
> > +            if(!hasNext()) { throw new NoSuchElementException(); }
> hasNext() here seems to throw an exception if the iterator is closed. Is this a reasonable thing to do? I would imagine that a better action would be to return false...not your fault, but an indication of the state of affairs. @alecw <https://github.com/alecw> I think you added these assertions, do you remember why?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/samtools/htsjdk/pull/480/files#r53542692>.
>